server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -636,8 +636,16 @@ Index resolveIfAbsent(DocWriteRequest<?> request) {
             Index concreteIndex = indices.get(request.index());
             if (concreteIndex == null) {
                 boolean includeDataStreams = request.opType() == DocWriteRequest.OpType.CREATE;
-                concreteIndex = indexNameExpressionResolver.concreteWriteIndex(state, request.indicesOptions(), request.indices()[0],
-                    false, includeDataStreams);
+                try {
+                    concreteIndex = indexNameExpressionResolver.concreteWriteIndex(state, request.indicesOptions(),
+                        request.indices()[0], false, includeDataStreams);
+                } catch (IndexNotFoundException e) {
+                    if (includeDataStreams == false && e.getMetadataKeys().contains("es.excluded_ds")) {
```

Comment:
Just a question -- I don't understand why `includeDataStreams == false` is here. I would have thought it would be the opposite if it were addressing the case indexing into data streams though it does seem to work according to the tests you added.

Reply:
I don't think it is needed (because `es.excluded_ds` is only set if `includeDataStreams` is `false`), but I like to keep it here to highlight that this check is only needed for op types other than create.  If `includeDataStreams` is `false` then `request.opType()` is not `CREATE`.